EE: Rework overclock and underclock function

ssakash opened the pull request:
**Summary of changes :-**

- don't classify blocks and use different scaled block cycles for each specific range, rather use a constant value for all the range (having lower blocks as an exception since they're very sensitive on these stuffs)

-  decrease the complexity on the code ( 71 lines -> 18 lines )

- Integrate ``eeScaleBlockCycles()`` and ``scaleBlockCycles_helper()`` to a single function

The first mentioned change actually seems to make the games run a lot better on overclock and underclock scenarios thanks to the removed classification of scalar_mid & scalar_ high stuffs. which probably caused some quirks due to dynamic scalar variable toggles (maybe ? )

The following code change needs reviews from @ramapcsx2  and @refractionpcsx2 

also for future plans,

- maybe move the cycle rate manipulation to some place rather than  ``iBranchTest()`` to have higher overclock range. ( higher underclock range is already possible )

* ~~invert the slider to have overclock on right side and underclock on left, lots of users were complaining about it. ( Need more users opinion before proceeding on this one)~~ Done :heavy_check_mark:

willkuer commented:
I am no friend of inverting the slider btw. To change an existing feature into the opposite is a usability crime. At least for a 'short' time. I dont see the point why overclocking needs to be to the right. Sure the emulated 'clock rate' increases with overclocking. But the time period between events is decreased and the unlocked fps are decreased (due to cpu limitations). To define one direction as the one and only truth is not possible here.

Actually it seems that the scalars translates into time period and the ee slider translatates monotonously into scalars. So the current slider is from a programmer point of view more correct than the reversed order.

i30817 commented:
good thing users aren't programmers then.

willkuer commented:
And what is the obvious effect of an increased speedhacks slider? Increased or decreased speed? Speed is btw. the title bar entry that compares gpu vsync events with expected vsync events.

If you want to invert the slider it should be named ee clockrate and put into the ee section of emulation settings. Because then it is imo not a speedhack anymore but an ee setting.

turtleli commented:
Should probably mention: lines of code != complexity. This code has quite a few readabilty issues IMO. I put them all here since there's quite a few and it prevents mass email spam.

```
bool lowcycles = !!(s_nBlockCycles <= 40);
```
The `!!` isn't necessary.
```
s8 cyclerate = (EmuConfig.Speedhacks.EECycleRate > 100 || EmuConfig.Speedhacks.EECycleRate < -2 ) ? 1 : EmuConfig.Speedhacks.EECycleRate; // use default cycle rate if users set unsupported values.
```
0, not 1. Also, swapping the two range checks would be more readable I think. IMO the below is even better:
```
s8 cyclerate = EmuConfig.Speedhacks.EECycleRate;
if (cyclerate < -2 || cyclerate > 100)
	cyclerate = 0;
```
```
else if (cyclerate < 0 && !lowcycles)
	scale_cycles = (-cyclerate) == 2 ? s_nBlockCycles >> 5 : (2 * s_nBlockCycles) >> 5; // should be fine since we only have two overclock modes currently.
```
`(-cyclerate) == 2` ? Same as `cyclerate == -2`?
`(2 * s_nBlockCycles) >> 5`? Same as `s_nBlockCycles >> 4`?
(And as a note, that leads to the expression `s_nBlockCycles >> (3 - cyclerate)`)
```
else if (!cyclerate || lowcycles)
	scale_cycles = s_nBlockCycles >> 3; // Default cyclerate 
```
`cyclerate == 0` is more appropriate here. Also, if you make this the initial if condition, you don't have to keep adding `&& !lowcycles` to every other else if and it's also more readable.

ssakash commented:
@turtleli 

> if (cyclerate < -2 || cyclerate > 100)
cyclerate = 0;

that's actually > 99 , also added that statement to the conditional check for scaled block cycles value rather than setting cycle rate to 0. 

like this : ``if (cyclerate == 0 || lowcycles || cyclerate > 99 || cyclerate < -2) scale_cycles = s_nBlockCycles >> 3;``

 > And as a note, that leads to the expression s_nBlockCycles >> (3 - cyclerate)

 Thanks for the equation !  didn't even notice this TBH, does make the code more readable. 

updated the code, feel free to do more reviews  :+1:

ssakash commented:
@willkuer  && @i30817 

I was planning to open a poll later on the forums for placement of the overclock / underclock slider values, don't worry this PR doesn't do the GUI change yet.  I'd probably only do that after a consensus among the users.

Blackbird88 commented:
I for one would like it to be inverted. Dolphin already has it like this and it makes lot more sense.

turtleli commented on pcsx2/gui/Panels/SpeedhacksPanel.cpp:
```diff
@@ -23,25 +23,26 @@ const wxChar* Panels::SpeedHacksPanel::GetEEcycleSliderMsg( int val )
 	switch( val )
 	{
+		switch( val )
+	{
```
I've been wondering for 2 days - compile checked?

ssakash replied:
Nope, that was done on Git online editor. should work fine now :) 

I'm not really sure what to do with it though , the users at poll thread seems to have lots of weird ideas @_@

ssakash commented:
@refractionpcsx2 

should we Invert the slider then ? it has most of the votes and lots of people said it was okay.

willkuer commented:
its funny that you ping refraction... the only dev who voted for inverting :P

ssakash commented:
he's the only dev who voted on the poll itself :p

willkuer commented on pcsx2/x86/ix86-32/iR5900-32.cpp:
```diff
@@ -1113,76 +1113,23 @@ void iFlushCall(int flushtype)
 {
-	// Note: s_nBlockCycles is 3 bit fixed point.  Divide by 8 when done!
-
-	// Let's not scale blocks under 5-ish cycles.  This fixes countless "problems"
```
I think I would not remove that comment. Since you follow it with bool lowcycles later.

ssakash replied:
yeah, that's important. I'll put it back when I get the time

willkuer commented:
Why have you actually removed low mid and high cycles? I guess they were initially put with some intention?

ssakash commented:
I mentioned it at my first comment 

> don't classify blocks and use different scaled block cycles for each specific range, rather use a constant value for all the range (having lower blocks as an exception since they're very sensitive on these stuffs)

mirhl commented:
What if we instead changed the "names"?
I mean: replace *cyclerate* with something that could put together the meaning of the slider inside the emulator with the actual "conceptual" expectation of people. 

I'd like to help more but I really wouldn't know all the coding tricks involved.

refractionpcsx2 commented:
@mirhl you mean like having "smoothness" or something? then you can have Faster <---------> Smoother

willkuer commented:
You said that you did it and that it runs better for you. Code complexity is better so nothing against that. The question is more why it was introduced in the first place and why you can remove it now.

mirhl commented:
Smoothness seems too stupid, really :p
Though, yeah. Isn't there any other idea that could be attached after EE?
First serious thing that comes to my mind is period (which is the inverse of frequency :s). 
But I'm pretty sure there has to be a better technical description

willkuer commented:
Actually i think the smoother <-> faster idea is the best we have so far. It is extremely intuitive.

refractionpcsx2 commented:
> First serious thing that comes to my mind is period (which is the inverse of frequency :s).

Yeah and users won't have a clue what you're talking about, you need either laymen's terms or buzzwords :P

ssakash commented:
> You said that you did it and that it runs better for you. Code complexity is better so nothing against that. The question is more why it was introduced in the first place and why you can remove it now.

the intention seems to be for scaling the blocks differently in case of these specific mid , high ranges. not really necessary to handle them differently and it doesn't provide uniform scaling of the blocks at all instances , so removed them to have it at a constant rate at all instances.

gregory38 commented:
Is it planned to have green text in EE cyclerate (vu stealing) for the default & safe value?

ssakash commented:
![main-emu](http://ultraimg.com/images/mainemu.png)

To be honest, Green doesn't suit the dialog background color. it's pretty much unreadable at first sight comparing to the VU slider red color text on the following screenshot.

refractionpcsx2 commented:
Or you could use a darker green, like RGB color #42AD3E

ssakash commented:
used the color you mentioned, thank you ![Rapid tables](http://www.rapidtables.com/web/color/RGB_Color.htm) !

![main](http://www.ultraimg.com/images/greenie.png)

better ? (or) should I try another one ? feel free to suggest any other better color.

refractionpcsx2 commented:
Hmm maybe still a bit bright, try #21A110

mirhl commented:
The color is nicer, but shouldn't we stick with black for neutral color?

What about green for underclocking (aka easier/faster emulation) and red for overclocking (aka more difficult/heavier emulation)?

refractionpcsx2 commented:
We want something that stands out a bit, to show this is a good thing :P Black will just blend in with all the other text on the page.

mirhl commented:
Yes, but my visual metaphor was more in line with green recalling the environment (aka less power hungry, aka less horsepower needed) and red recalling all those overclocked and aggressively marketed products you see around. 

This way I saw there could be less room for misunderstandings.

willkuer commented:
But green is always better, or? Contrary ee underclock - especially maxed - might not be the best. I would agree with green = default.

gregory38 commented:
For me default is safest. Speed hacks are only a necessary for limited computer

mirhl commented:
@gregory38 Technically if we consider overclocking EE it's not only for those cases.

gregory38 commented:
Honestly I don't care but one value must be the default to ease configuration. All red (or black) is just too complicated.

wareya commented:
Since both directions have the potential to change game behavior in unexpected ways, then if anything is "green", only the default should be "green". So either the default settings should be singled out as being safe by making them green, or nothing should be singled out.

Making one way green and one way red is silly, because both change the game behavior away from the pure form. .....Not like PCSX2 is perfectly cycle accurate, but....

ssakash commented:
okay, here's how it looks with the #42AD3E color on the texts. doesn't really look much different but it's readable. what do you think @refractionpcsx2  ? 

![image](http://ultraimg.com/images/newemu.png)

willkuer commented:
Its still worse to read than simple black. Should the slider setting indicator (single number below slider) also be colored?

ssakash commented:
I used a slightly darker green and it looks quite readable now.
![greenie](http://ultraimg.com/images/greens.png)

> Should the slider setting indicator (single number below slider) also be colored?

In my opinion, nope. It has always been black and should stay the same way , a single colored number doesn't matter much.

avih commented:
> The first mentioned change actually seems to make the games run a lot better on overclock and underclock scenarios thanks to the removed classification of scalar_mid & scalar_ high stuffs. which probably caused some quirks due to dynamic scalar variable toggles (maybe ? )

I didn't test PCSX2 for a while, and recently noticed that ICO now plays quite badly in preset 3 ("balanced"), while in the past preset 3 was great for ICO. Specifically, the audio has a lot of skips and slow downs (while the title shows 100% speed). It works now fine with preset 2 ("Safe").

I bisected and found out it's due to the change at this PR.

When I looked at the new calculation (which is the same calculation we have today, except the reversal of the slider), I noticed that one click left of the default (which was one click to the right in the past) which was using low/mid/high of 5/7/5 respectively, now works as if all three are 7, and two clicks which had low/mid/high of 7/9/7 now works as if all are 9.

While it's OK to simplify code, in this case it actually makes these two slider positions, and especially the 1-click position quite a bit more extreme than it used to be.

Some calculations:
- I tested several games and counted how many times the low/high were used compared to how many times the mid value got used (for both click positions the low/high had the same scaler, and the mid scaler was higher). Turns out that pretty much all games end up with a ratio of about 6, i.e. the low/high were used x6 the number of times mid was used.
- For the 1 click position, which had low/mid/high of 5/7/5, on average it applied a scaler of (6 x 5 + 1 x 7) / 7 = 5.3, while your new calculation always uses a scaler of 7. In other words, you increased the average scaler by ~33%.
- With a similar calculation, you increased the scaler for the 2-clicks position from average of 7.3 to always 9, which is an increase of almost 25%.

Effectively, you took the most extreme scaler you could use in the past with the slider (2 clicks, 7.3 scaler on average), and made it the new value of the first click (7 on average).

Now, supposedly 2 clicks position was relatively extreme (and indeed it affected many games badly), and the 1 click position was mild, and it's used at the "balanced" preset (and also at the "aggressive" preset). But now, that extreme value is the new minimum scaler which can be applied, and also the value which the balanced preset uses. This is not great.

Unless there are any objections, I'm going to restore these two slider positions to have the same effect they had before this PR landed (without touching the calculations of the other slider positions or calculations of values beyond the slider range if set via the ini file).

ssakash commented:
@avih 

I'm fine with it as long as you open a PR with ''Needs testing`` label. IIRC, the increased average worked better with some games compared to the older behavior.

It'd be nice to compare the older behaviour with the new one on even more testcases. (AKA games)

avih commented:
> IIRC, the increased average worked better with some games compared to the older behavior.

It would help if you listed those games, since so far you haven't named even one.

ssakash commented:
> It would help if you listed those games, since so far you haven't named even one 

I would have mentioned if I had remembered them, I coded this PR last year so no clue :P 

I _think_ ``Naruto Ultimate Ninja`` and ``God of War`` series noticed an improvement with the new change.

avih commented:
> I would have mentioned if I had remembered them, I coded this PR last year so no clue :P

Well, you based this change on the notion that it improved things, but never actually listed any games which you tested.

While at it, batman begins also doesn't boot with the balanced preset now, again, because the mildest underclock we can apply with these values is still too extreme for it. Same as Ico.

refractionpcsx2 commented:
I think the performance of the hack should be based on the average across many games rather than a select couple.  After all, it is a hack and it will break things, so where it shows an average improvement is the best place to put it.

ssakash commented:
@avih

Making the slider values more granular would also be helpful to find the perfect scale block size for specific games. (Avoiding ICO , Batman issue by using lower values)

avih commented:
> I think the performance of the hack should be based on the average across many games

Of course, which the original calculation had as it was refined over many years. But the new value is there because one guy says "The first mentioned change actually seems to make the games run a lot better" without actually providing any kind of data.

avih commented:
> For the 1 click position, which had low/mid/high of 5/7/5, on average it applied a scaler of (6 x 5 + 1 x 7) / 7 = 5.3, while your new calculation always uses a scaler of 7. In other words, you increased the average scaler by ~33%.

Actually, I underestimated the change.

Turns out that in practice, before the change in this PR, the mildest notch was about 25% reduction, and afterwards it became 40% reduction. That's a really big reduction for the mildest underclock value.

I measured the change in speed for all the current slider positions and updated their texts to be more accurate. The measurement code and the updated texts are at 67dc3eef4b97be26a55d4afb4f3fa63697791f2e .

I then also added back the original mild value to slider position -1, and so the balanced preset now uses it too. The other two slider positions are still there too, but pushed down one notch. That's at 460b7be47abb8db7d414ded175da50094b353b54 .

refractionpcsx2 commented:
@avih those sliders no longer make sense, now you have to put the slider to the extreme for the smaller underclock and next to zero for the biggest. If you're going to do that you might as well swap the values around, unless you put them in the wrong way around.

avih commented:
> @avih those sliders no longer make sense, now you have to put the slider to the extreme for the smaller underclock and next to zero for the biggest. If you're going to do that you might as well swap the values around, unless you put them in the wrong way around.

I don't think so, unless I missed something?

It was: `A   B  0  C  D` where B is the mildest underclock and C is the mildest overclock.

Now it's `A  B  X  0  C  D` where the only change is the new X which is a milder underclock that B.

Did I miss anything?

refractionpcsx2 commented:
In here you swapped things around

https://github.com/PCSX2/pcsx2/commit/67dc3eef4b97be26a55d4afb4f3fa63697791f2e

Then you made it worse when you added another one :P

avih commented:
> In here you swapped things around
> 
> 67dc3ee
> 
> Then you made it worse when you added another one :P

What? this commit only updates the text to be more correct...

refractionpcsx2 commented:
Then from a user experience point of view the scalings should be swapped around so it goes from smallest closest to 0 to biggest furthest away.  At the moment it's confusing.

avih commented:
OK, now I see your point. However, describing a change in percentage is highly misleading. for instance -50% is really not symmetric to +50%. The change just make it shows the value we're setting it to, rather than the change from the middle position.

And people can relate better, IMO, to 20% speed vs 100% speed vs 200% speed, rather than calling this exact triplet "-80%, 0, +100%".

Regardless, if you think that "-50%, -40%, -25%, (0), +80%, +200%" is better than "50%, 60%, 75%, (100%), 180%, 300%" (these are the two representations of exactly the same set of measured values), I'm absolutely fine with your way too.

refractionpcsx2 commented:
I see what you are getting at now and i did slightly misread it. However I personally feel it makes more sense (for the user and me, IMO ofc :P ) to say "this value is 20% faster than normal" etc, but more elegantly.

avih commented:
> However I personally feel it makes more sense ... to say "this value is 20% faster than normal"

Like I said, I'm fine with whatever more users would find easier to grasp.

I'll give people a chance to voice their opinion on this subject, and if after a day or two no one cares or most people prefers it your way, I'll restore the texts it to how they were before.


> But the values should be more uniform

I don't have a strong opinion on this one, and I didn't modify the effect which each slider position has on PCSX2.

I only measured what's the actual change in practice, rather than guesses of "increase of 33%" etc, and updated the text to be more accurate.

Fine tuning presets is black magic, and while I know that some presets work very well (like the mildest underclock before this PR, and which was reintroduced again recently), I don't claim to know what are the best values here.

avih commented:
> with a maximum percentage of say 75% clock either way of normal

You do notice you fall into the "change in percentage" trap which my modified texts try to prevent, right?

"clock speed reduced by 75%" is 4 times slower, and "clock speed increased by 75%" is 1.75 times faster.

avih commented:
@refractionpcsx2 seems no one cares :)

Just give me the word, and I'll update the texts to use "by NN%" again (but with the corrected values).

Waiting for your confirmation before I update the texts.

refractionpcsx2 commented:
Yeah just do whatever :P People are only going to put the slider up to max anyways lol

avih commented:
K, so I left it as is for now. If we get any evidence users like it less, or if you change your mind, or if more devs prefer it as it was before, I'll have no objection to revert it.

gregory38 commented:
A silly question, but are percentage really necessary. People suck in math, and nobody could translate N% to my game will run M% faster/slower.

avih commented:
They're definitely not "really necessary", but I think we do want to give some kind of indication to the amount of over/under clock we're applying, instead of saying "default", "a little underclock" "a bit more underclock", etc :)

How would you choose to describe the different slider positions?

gregory38 commented:
Well, it is done for VU cycle stealing, nobody complains about it.

willkuer commented:
Probably my opinion barely matters but I highly agree with avih regarding the absolute differences in %. They are really misleading and not what you actually aim for. Absolute values as suggested makes really more sense.

However maybe one can do a scaling like the gsdx resolution (native,2xnative,3xnative) so maybe something like:
base frequency: stock
overclocking: stock x 1.2, stock x 1.5, stock x 2
underclocking: stock / 1.2, stock / 1.5, stock / 2 (or stock : 1.2, stock : 1.5, stock : 2)

I have the feeling if you tell somebody he is overclocking the EE to twice its stock or he is underclocking the clockrate to half of it he will instantly understand.

avih commented:
I agree that a direct factor ("x 1.5") could help, but then you need to use similar terminology for both underclock and overclock, and while I think "x2" or "x1.5" is somewhat useful, the other side of "/ 1.5" is not very intuitive IMO.

But yeah, if we can find a way to describe it using a factor while keeping it reasonably intuitive for users (and to us!), I think it would be better than percentage (better than both absolute and relative percentage).
